### PR TITLE
Add ERC721TransferEnforcer

### PR DESCRIPTION
**Pull request by danfinlay**
### **What?**

- Adds an ability to share ERC721 transfer rights

### **Why?**

- Because it's a basic and essential, frequently shared ability.

### **How?**

- With AI and some elbow grease.

### CODE REVIEW

File: test/enforcers/ERC721TransferEnforcer.t.sol [hunk 1]
```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: MIT AND Apache-2.0
+pragma solidity 0.8.23;
+
+import "forge-std/Test.sol";
```
Comment (hanzel98) [inline test/enforcers/ERC721TransferEnforcer.t.sol:4]: We can delete this import

**Comment by Maliksb11** [comment]
Ohh 😯

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        // Decode the remaining callData into NFT transfer parameters
+        if (callData_.length < 100) {
+            revert("ERC721TransferEnforcer:invalid-calldata-length");
+        }
+        
+        address from;
+        address to;
+        uint256 transferTokenId;
+        
+        if (callData_.length < 100) {
```
Comment (dylandesrosier) [inline src/enforcers/ERC721TransferEnforcer.sol:33]: Duplicate of line 25 check

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            revert("ERC721TransferEnforcer:invalid-calldata-length");
+        }
+        
+        (from, to, transferTokenId) = abi.decode(callData_[4:], (address, address, uint256));
+        
+        if (from == address(0) || to == address(0)) {
+            revert("ERC721TransferEnforcer:invalid-address");
+        }
+
+        if (from == address(0) || to == address(0)) {
```
Comment (dylandesrosier) [inline src/enforcers/ERC721TransferEnforcer.sol:43]: Duplicate of line 39 check

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        if (target != permittedContract) {
+            revert("ERC721TransferEnforcer:unauthorized-contract-target");
+        } else if (selector != IERC721.transferFrom.selector) {
+            revert("ERC721TransferEnforcer:unauthorized-selector");
+        } else if (transferTokenId != permittedTokenId) {
+            revert("ERC721TransferEnforcer:unauthorized-token-id");
+        }
+    }
+
+    function afterHook(
```
Comment (dylandesrosier) [inline src/enforcers/ERC721TransferEnforcer.sol:55]: No need to override hooks with nothing if you're extending from `CaveatEnforcer` as it provides empty defaults for all of the hooks.

**Comment by hanzel98** [comment]
We need the new deployment script

File: src/enforcers/ERC721TransferEnforcer.sol
Comment (hanzel98) [inline src/enforcers/ERC721TransferEnforcer.sol:1]: The variables should use underscore
```solidity
    function beforeHook(
        bytes calldata _terms,
        bytes calldata,
        ModeCode,
        bytes calldata _executionCallData,
        bytes32,
        address,
        address
    )
        public
        virtual
        override
    {
        (address permittedContract_, uint256 permittedTokenId_) = getTermsInfo(_terms);
        (address target_,, bytes calldata callData_) = ExecutionLib.decodeSingle(_executionCallData);
        bytes4 selector_ = bytes4(callData_[0:4]);

        // Decode the remaining callData into NFT transfer parameters
        if (callData_.length < 100) {
            revert("ERC721TransferEnforcer:invalid-calldata-length");
        }

        address _from;
        address _to;
        uint256 _transferTokenId;
        (_from, _to, _transferTokenId) = abi.decode(callData_[4:], (address, address, uint256));

        if (_from == address(0) || _to == address(0)) {
            revert("ERC721TransferEnforcer:invalid-address");
        }

        if (target_ != permittedContract_) {
            revert("ERC721TransferEnforcer:unauthorized-contract-target");
        } else if (selector_ != IERC721.transferFrom.selector) {
            revert("ERC721TransferEnforcer:unauthorized-selector");
        } else if (_transferTokenId != permittedTokenId_) {
            revert("ERC721TransferEnforcer:unauthorized-token-id");
        }
    }

    function getTermsInfo(bytes calldata _terms) public pure returns (address permittedContract_, uint256 permittedTokenId_) {
        if (_terms.length != 52) revert("ERC721TransferEnforcer:invalid-terms-length");
        permittedContract_ = address(bytes20(_terms[:20]));
        permittedTokenId_ = uint256(bytes32(_terms[20:]));
    }
  ```

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        bytes32,
+        address,
+        address
+    )
+        public
+        virtual
+        override
+    {
+        (address permittedContract, uint256 permittedTokenId) = getTermsInfo(_terms);
+        (address target,, bytes calldata callData_) = ExecutionLib.decodeSingle(executionCallData);
```
Comment (hanzel98) [inline src/enforcers/ERC721TransferEnforcer.sol:26]: We are assuming this is a single execution, we can use this modifier `onlySingleExecutionMode`

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        // Decode the remaining callData into NFT transfer parameters
+        if (callData_.length < 100) {
+            revert("ERC721TransferEnforcer:invalid-calldata-length");
+        }
+
+        address from_;
+        address to_;
+        uint256 transferTokenId_;
+        (from_, to_, transferTokenId_) = abi.decode(callData_[4:], (address, address, uint256));
```
Comment (hanzel98) [inline src/enforcers/ERC721TransferEnforcer.sol:36]:
```suggestion
 (address _from, address _to, uint256 _transferTokenId) = abi.decode(callData_[4:], (address, address, uint256));
```

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        (address from_, address to_, uint256 transferTokenId_) = abi.decode(callData_[4:], (address, address, uint256));
+
+        if (from_ == address(0) || to_ == address(0)) {
+            revert("ERC721TransferEnforcer:invalid-address");
+        }
+
+        if (target_ != permittedContract_) {
+            revert("ERC721TransferEnforcer:unauthorized-contract-target");
+        } else if (selector_ != IERC721.transferFrom.selector) {
+            revert("ERC721TransferEnforcer:unauthorized-selector");
```
Comment (hanzel98) [inline src/enforcers/ERC721TransferEnforcer.sol:54]: I can't see a test for this error

File: src/enforcers/ERC721TransferEnforcer.sol [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ * @title ERC721TransferEnforcer
+ * @notice This enforcer restricts the action of a UserOp to the transfer of a specific ERC721 token.
+ */
+contract ERC721TransferEnforcer is CaveatEnforcer {
+    /**
+     * @notice Enforces that the contract and tokenId are permitted for transfer
+     * @param _terms the encoded contract and tokenId
+     * @param _mode the execution mode of the transaction
+     * @param _executionCallData the call data of the transferFrom call
+     */
```
Comment (hanzel98) [inline src/enforcers/ERC721TransferEnforcer.sol:19]: In other enforcers we use this natspec to detail how terms are formed.

**Review by hanzel98 — APPROVED** [review]

**Pull request merged**